[InstCombine] Simplify phi using KnownBits of condition

andjo403 opened the pull request:
Simplify the arms of a phi based on the KnownBits implied by the condition for the predecessor basic block.

I noticed regression of the form in the proof for a PR that I work on, I Do not know how if possible to make the proof generic.

proof: https://alive2.llvm.org/ce/z/Xh7CQb

dtcxzyw commented:
Compile-time overhead: +~0.2%

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -840,17 +840,28 @@ void llvm::computeKnownBitsFromContext(const Value *V, KnownBits &Known,
-        computeKnownBitsFromCond(V, BI->getCondition(), Known, Depth, Q,
-                                 /*Invert*/ false);
+    const BasicBlock *CxtIBB = Q.CxtI->getParent();
+    if (isa<PHINode>(Q.CxtI))
```
We don't need to specially handle the PHI node here. You should pass a more precise CxtI (See the comments below).

dtcxzyw replied:
I see. We need the edge condition `Pred->BB` to improve the result.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombinePHI.cpp:
```diff
@@ -1625,6 +1625,25 @@ Instruction *InstCombinerImpl::visitPHINode(PHINode &PN) {
 
+  if (PN.getType()->isIntegerTy()) {
+    bool MadeChange = false;
+    SimplifyQuery Q = SQ.getWithInstruction(&PN);
+    for (unsigned I = 0, E = PN.getNumIncomingValues(); I != E; ++I) {
+      Value *V = PN.getIncomingValue(I);
```
```suggestion
    for (unsigned I = 0, E = PN.getNumIncomingValues(); I != E; ++I) {
      SimplifyQuery Q = SQ.getWithInstruction(PN.getIncomingBlock(I)->getTerminator());
      Value *V = PN.getIncomingValue(I);
```

andjo403 commented on llvm/lib/Transforms/InstCombine/InstCombinePHI.cpp:
```diff
@@ -1625,6 +1625,25 @@ Instruction *InstCombinerImpl::visitPHINode(PHINode &PN) {
+      Value *V = PN.getIncomingValue(I);
+      if (isa<ConstantInt>(V))
+        continue;
+
```
> Compile-time overhead: +~0.2%

maybe having a check like this can help the compile time regression:

```suggestion
      if (auto *BI =
              dyn_cast<BranchInst>(PN.getIncomingBlock(I)->getTerminator()))
        if (BI->isUnconditional())
          continue;

```
think that it only is for this that this simplification is needed

andjo403 replied:
hmm one test failed when I tried this so not only for conditional branches that this simplify

dtcxzyw commented:
> Compile-time overhead: +~0.2%

11728 files changed, 23367619 insertions(+), 24618099 deletions(-)
TBH this patch leads to many attractive optimizations on real-world programs :) I think the compile-time overhead is acceptable. Can we further reduce the overhead by limiting the recursive depth?

However, specially-handling the PHI node in `computeKnownBitsFromContext` is still an ugly way. Can you try `CondContext`?

davemgreen commented on llvm/test/Transforms/LoopVectorize/AArch64/intrinsiccost.ll:
```diff
@@ -7,10 +7,6 @@ target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128"
 target triple = "aarch64--linux-gnu"
 
 ; CHECK-COST-LABEL: sadd
-; CHECK-COST: Found an estimated cost of 6 for VF 1 For instruction:   %1 = tail call i16 @llvm.sadd.sat.i16(i16 %0, i16 %offset)
```
Why are these removed?

andjo403 replied:
due to the misscompile

davemgreen commented on llvm/test/Transforms/LoopVectorize/ARM/mve-reductions.ll:
```diff
@@ -70,11 +70,10 @@ define i64 @add_i32_i64(ptr nocapture readonly %x, i32 %n) #0 {
 ; CHECK-NEXT:    br label [[FOR_BODY:%.*]]
 ; CHECK:       for.body:
-; CHECK-NEXT:    [[I_08:%.*]] = phi i32 [ [[INC:%.*]], [[FOR_BODY]] ], [ [[BC_RESUME_VAL]], [[SCALAR_PH]] ]
+; CHECK-NEXT:    [[I_08:%.*]] = phi i32 [ [[INC:%.*]], [[FOR_BODY]] ], [ 0, [[SCALAR_PH]] ]
```
Are you sure this is correct, if it is coming from middle.block via the vectorized body?

dtcxzyw replied:
Looks like this patch causes some miscompilation: https://github.com/dtcxzyw/llvm-fuzz-service/issues/65

andjo403 replied:
Yes sorry about that I do not know how I missed this it is fixed now

dtcxzyw reviewed (changes requested):
Miscompilation reproducer: https://alive2.llvm.org/ce/z/X5EbE5
```
; bin/opt -passes=instcombine test.ll -S
define i16 @src(i16 %ui1) {
entry:
  %cmp = icmp eq i16 %ui1, 0
  br i1 %cmp, label %cond.end, label %cond.false

cond.false:                                       ; preds = %entry
  br label %cond.end

cond.end:                                         ; preds = %cond.false, %entry
  %cond.in = phi i16 [ %ui1, %cond.false ], [ 0, %entry ]
  ret i16 %cond.in
}
```
```
define i16 @tgt(i16 %ui1) {
entry:
  %cmp = icmp eq i16 %ui1, 0
  br i1 %cmp, label %cond.end, label %cond.false

cond.false:                                       ; preds = %entry
  br label %cond.end

cond.end:
  ret i16 0
}
```

andjo403 committed:
[InstCombine] Simplify phi using KnownBits of condition

Simplify the arms of a phi based on the KnownBits implied
by the condition for the predecessor basic block.

andjo403 commented:
> However, specially-handling the PHI node in `computeKnownBitsFromContext` is still an ugly way. Can you try `CondContext`?

not only was it ugly it was the cause of the Miscompilation so fixed now

andjo403 commented on llvm/lib/Transforms/InstCombine/InstCombinePHI.cpp:
```diff
@@ -1625,6 +1625,44 @@ Instruction *InstCombinerImpl::visitPHINode(PHINode &PN) {
+      if (BI && BI->isConditional() && llvm::is_contained(BRs, BI)) {
+        CondContext CC(BI->getCondition());
+        CC.AffectedValues.insert(V);
+        if (BI->getSuccessor(1) == PNBB)
```
I assumed that the two branched did not have the same label but found this https://github.com/llvm/llvm-project/blob/72436b37bf4203ee43395c65cc179dc573f79251/llvm/lib/Analysis/ValueTracking.cpp#L1735-L1736 

But I'm not able to create a test that have the same label for the two branches https://godbolt.org/z/avb161nYe
so is this a case that I shall handle?

nikic replied:
https://godbolt.org/z/vacTYPoha is the way to spell it.

andjo403 replied:
Thanks for the help with how to make the test.
I do not think that it needs to be handled here as the condition in the branch will be replaced by a constant so it will not be in the DomConditionCache.
see https://github.com/llvm/llvm-project/blob/72436b37bf4203ee43395c65cc179dc573f79251/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L3787-L3791

andjo403 commented:
@dtcxzyw can you re-tigger the benchmarks again, 
think that the compile time shall be better as only values in the DomConditionCache will be simplified and
 will also not simplify as mush now when the miscompilation is solved.

dtcxzyw commented:
> @dtcxzyw can you re-tigger the benchmarks again, think that the compile time shall be better as only values in the DomConditionCache will be simplified and will also not simplify as mush now when the miscompilation is solved.

Feel free to submit task here: https://github.com/dtcxzyw/llvm-opt-benchmark/issues/1312

andjo403 commented:
> Compile-time overhead: +~0.2%

The compile time looks mush better now even an overall improvement 
Compile-time overhead: -~0.02%

andjo403 commented:
found that I have lost the fold that I was trying to solve so not ready for review

andjo403 commented:
closing as this implementation is wrong and it also feels like this partly reverts the change from https://github.com/llvm/llvm-project/commit/f1106ef6c9d14d5b516ec352279aeee8f9d12818
